packages/data-context/src/sources/ProjectDataSource.ts

```diff
@@ -335,17 +336,40 @@ export class ProjectDataSource {
     // We respond to all changes to the project's filesystem when
     // files or directories are added and removed that are not explicitly
     // ignored by config
-    this._specWatcher = chokidar.watch('.', {
-      ignoreInitial: true,
-      ignorePermissionErrors: true,
-      cwd: projectRoot,
-      ignored: ['**/node_modules/**', ...excludeSpecPattern, ...additionalIgnorePattern],
+    this._specWatcher = this._makeSpecWatcher({
+      projectRoot,
+      specPattern,
+      excludeSpecPattern,
+      additionalIgnorePattern,
     })
 
     // the 'all' event includes: add, addDir, change, unlink, unlinkDir
     this._specWatcher.on('all', onProjectFileSystemChange)
   }
 
+  _makeSpecWatcher ({ projectRoot, specPattern, excludeSpecPattern, additionalIgnorePattern }: { projectRoot: string, excludeSpecPattern: string[], additionalIgnorePattern: string[], specPattern: string[] }) {
+    return chokidar.watch('.', {
+      ignoreInitial: true,
+      ignorePermissionErrors: true,
+      cwd: projectRoot,
+      ignored: ['**/node_modules/**', ...excludeSpecPattern, ...additionalIgnorePattern, (file: string) => {
+        // don't ignore things that look like directories,
+        // using statSync, falling back to whether it has an extension name
+        // if that were to throw
+        try {
+          if (fs.statSync(file).isDirectory()) {
```

Comment:
`ignored`'s second argument has `Fs.Stats`, so you don't need to do a synchronous `stat` here: https://github.com/paulmillr/chokidar#path-filtering

Also, avoid doing synchronous FS operations, it blocks the entire server when you do this.

Reply:
Updating